allennlp/modules/matrix_attention.py

```diff
@@ -0,0 +1,58 @@
+import torch
+from overrides import overrides
+
+from allennlp.common import Params
+from allennlp.modules.similarity_function import SimilarityFunction
+
+
+class MatrixAttention(torch.nn.Module):
+    '''
+    This ``Module`` takes two matrices as input and returns a matrix of attentions.
+
+    We compute the similarity between each row in each matrix and return unnormalized similarity
+    scores.  Because these scores are unnormalized, we don't take a mask as input; it's up to the
+    caller to deal with masking properly when this output is used.
+
+    By default similarity is computed with a dot product, but you can alternatively use a
+    parameterized similarity function if you wish.
+
+    This is largely similar to using ``TimeDistributed(Attention)``, except the result is
+    unnormalized.  You should use this instead of ``TimeDistributed(Attention)`` if you want to
+    compute multiple normalizations of the attention matrix.
+
+    Input:
+        - matrix_1: ``(batch_size, num_rows_1, embedding_dim)``
+        - matrix_2: ``(batch_size, num_rows_2, embedding_dim)``
+
+    Output:
+        - ``(batch_size, num_rows_1, num_rows_2)``
+
+    Parameters
+    ----------
+    similarity_function: ``SimilarityFunction``, optional (default=``DotProductSimilarity``)
+        The similarity function to use when computing the attention.
+    '''
+    def __init__(self, similarity_function: SimilarityFunction = None) -> None:
+        super(MatrixAttention, self).__init__()
+
+        # This import is here to avoid circular import problems with the registry...
+        from allennlp.modules.similarity_functions.dot_product import DotProductSimilarity
```

Comment:
We need to have a better solution than this. This is unsustainable and confusing for contributors in the future - what that solution is, I am not exactly sure though ...